tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/iterator/FedXZeroLengthPathIteration.java

```diff
@@ -0,0 +1,220 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *******************************************************************************/
+package org.eclipse.rdf4j.federated.evaluation.iterator;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import org.eclipse.rdf4j.common.iteration.CloseableIteration;
+import org.eclipse.rdf4j.common.iteration.LookAheadIteration;
+import org.eclipse.rdf4j.federated.algebra.EmptyStatementPattern;
+import org.eclipse.rdf4j.federated.algebra.ExclusiveStatement;
+import org.eclipse.rdf4j.federated.algebra.FedXZeroLengthPath;
+import org.eclipse.rdf4j.federated.algebra.StatementSource;
+import org.eclipse.rdf4j.federated.algebra.StatementSourcePattern;
+import org.eclipse.rdf4j.federated.structures.QueryInfo;
+import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.Value;
+import org.eclipse.rdf4j.model.impl.SimpleValueFactory;
+import org.eclipse.rdf4j.query.BindingSet;
+import org.eclipse.rdf4j.query.MutableBindingSet;
+import org.eclipse.rdf4j.query.QueryEvaluationException;
+import org.eclipse.rdf4j.query.algebra.StatementPattern;
+import org.eclipse.rdf4j.query.algebra.StatementPattern.Scope;
+import org.eclipse.rdf4j.query.algebra.TupleExpr;
+import org.eclipse.rdf4j.query.algebra.Var;
+import org.eclipse.rdf4j.query.algebra.evaluation.EvaluationStrategy;
+import org.eclipse.rdf4j.query.algebra.evaluation.QueryBindingSet;
+import org.eclipse.rdf4j.query.algebra.evaluation.QueryEvaluationStep;
+import org.eclipse.rdf4j.query.algebra.evaluation.impl.QueryEvaluationContext;
+import org.eclipse.rdf4j.query.algebra.evaluation.iterator.CrossProductIteration;
+import org.eclipse.rdf4j.query.algebra.evaluation.iterator.ZeroLengthPathIteration;
+
+/**
+ * An iteration to evaluated {@link FedXZeroLengthPath}
+ *
+ * @see ZeroLengthPathIteration
+ */
+public class FedXZeroLengthPathIteration extends LookAheadIteration<BindingSet, QueryEvaluationException> {
+
+	/*
+	 * IMPL NOTE:
+	 *
+	 * This is technically almost a 1:1 copy of
+	 * org.eclipse.rdf4j.query.algebra.evaluation.iterator.ZeroLengthPathIteration Reusing or extending
+	 * ZeroLengthPathIteration requires refactoring in its constructor initialization.
+	 *
+	 * The main difference is in keeping track of QueryInfo and statement sources to be used in the precompiled
+	 * statement. Additionally the variable names for anon vars are renamed
+	 */
+
+	private static final Literal OBJECT = SimpleValueFactory.getInstance().createLiteral("object");
+
+	private static final Literal SUBJECT = SimpleValueFactory.getInstance().createLiteral("subject");
+
+	// Note: in contrast to the original zero length path iteration we use "_" instead of "-"
+	// as we need variable names valid in SELECT queries
+
+	private static final String ANON_SUBJECT_VAR = "zero_length_internal_start";
```

Comment:
here I had to use "_" instead of "-" as in the federation I generate SELECT queries to be sent to the members. "-" is not a valid character in variable names

Reply:
We should probably change that in the original implementation too. I'll take a look.